sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,318 @@
+const IMDS_SESSION_TOKEN_ENDPOINT: &str = "http://169.254.169.254/latest/api/token";
+
+#[derive(Serialize)]
+struct UserData {
+    ssh: Ssh,
+}
+
+#[derive(Serialize)]
+struct Ssh {
+    authorized_keys: Vec<String>,
+}
+
+/// Returns the value of a metadata key for a given data key, erroring if the value is not a
+/// string or is empty.
+async fn get_metadata(key: &str, meta: &str) -> Result<String> {
+    let uri = &format!("{}{}?keys={}", API_METADATA_URI_BASE, meta, key);
+    let method = "GET";
+    let (code, response_body) = apiclient::raw_request(DEFAULT_API_SOCKET, &uri, method, None)
+        .await
+        .context(error::APIRequest { method, uri })?;
+    ensure!(
+        code.is_success(),
+        error::Response {
+            method,
+            uri,
+            code,
+            response_body
+        }
+    );
+
+    // Metadata responses are of the form `{"data_key": METADATA}` so we pull out the value.
+    let mut response_map: HashMap<String, serde_json::Value> =
+        serde_json::from_str(&response_body).context(error::DeserializeJson)?;
+    let response_val = response_map
+        .remove(key)
+        .context(error::MissingMetadata { meta, key })?;
+
+    // Ensure it's a non-empty string
+    let response_str = response_val
+        .as_str()
+        .with_context(|| error::MetadataWrongType {
+            meta,
+            expected: "string",
+            value: response_val.to_string(),
+        })?;
+    let metadata = if !response_str.is_empty() {
+        response_str.to_string();
+    } else {
+        String::new();
+    };
+    Ok(metadata)
+}
+
+fn get_text_from_imds(client: &Client, uri: &str, session_token: &str) -> Result<String> {
+    client
+        .get(uri)
+        .header("X-aws-ec2-metadata-token", session_token)
+        .send()
+        .context(error::ImdsRequest { method: "GET", uri })?
+        .error_for_status()
+        .context(error::ImdsResponse { uri })?
+        .text()
+        .context(error::ImdsText { uri })
+}
+
+/// Returns a list of public keys.
+fn get_public_keys_from_imds() -> Result<Vec<String>> {
+    info!("Fetching IMDS session token");
+    let client = Client::new();
+    // Use IMDSv2 for accessing instance metadata.
+    let uri = IMDS_SESSION_TOKEN_ENDPOINT;
+    let imds_session_token = client
+        .put(uri)
+        .header("X-aws-ec2-metadata-token-ttl-seconds", "60")
+        .send()
+        .context(error::ImdsRequest { method: "PUT", uri })?
+        .error_for_status()
+        .context(error::ImdsResponse { uri })?
+        .text()
+        .context(error::ImdsText { uri })?;
+
+    info!("Retrieving list of available public keys from IMDS");
+    let public_key_list =
+        get_text_from_imds(&client, IMDS_PUBLIC_KEY_BASE_URL, &imds_session_token)?;
+    debug!("available public keys '{}'", &public_key_list);
+
+    info!("Generating endpoint urls to retrieve text of available public keys");
```

Comment:
This can probably be `debug!()`.  It's an implementation detail related to fetching public keys.